sklearn/ensemble/gradient_boosting.py

```diff
@@ -1073,7 +1114,31 @@ def _init_decision_function(self, X):
         if X.shape[1] != self.n_features:
             raise ValueError("X.shape[1] should be {0:d}, not {1:d}.".format(
                 self.n_features, X.shape[1]))
-        score = self.init_.predict(X).astype(np.float64)
+            # init predictions
+
+        if hasattr(self.init_, 'predict_proba'):
+            eps = np.finfo(X.dtype).eps
+            score = self.init_.predict_proba(X) + eps
+            if self.n_classes == 2:
+                score = np.log(score[:,1] / score[:,0])
```

Comment:
The binomial deviance indeed expects a predict function with values in `[-inf, +inf]` when computed by the basic `LogOdds` estimator instead of `[0, 1]` if we used the predicted proba for the positive class.

We should definitely gradient-check (and check that the computed values of the loss functions match the descriptions of the docstrings) for all those loss functions but this should be done in another PR.

+1 for keeping `np.log(score[:, 1] / score[:, 0])` for now.